### ISSUE TITLE
Condition on non-zero length of string literal doesn't short circuit `and` statements

### ISSUE THREAD

**Issue by kokes** [astral-sh/ty#1983]
### Summary

Code that checks `if len(value) and value[0] == ' '` should work just fine, but ty reports it as `index-out-of-bounds` in some cases (when the value is `foo | Literal[""]`).

Minimal code

```py
lines: list[str] = ["foo", "bar", "baz"]
for line in lines:
    if not line:
        continue

    value = line if len(line) < 3 else ""
    if len(value) and value[0] == " ":
        value = value[1:]
```

uvx ty check:
```
error[index-out-of-bounds]: Index 0 is out of bounds for string `Literal[""]` with length 0
 --> foo.py:7:23
  |
6 |     value = line if len(line) < 3 else ""
7 |     if len(value) and value[0] == " ":
  |                       ^^^^^
8 |         value = value[1:]
  |
info: rule `index-out-of-bounds` is enabled by default

Found 1 diagnostic
```

Changing the `if` condition to `if value and value[0]` makes it go away as well. Removing `if not line` (and turning `line` to a plain `str` also makes this go away. Seems like `ty` doesn't see through `Literal` and doesn't infer that `len(Literal[""])` is `AlwaysFalsy` (or something along those lines).

### Version

ty 0.0.2 (42835578d 2025-12-16)
